File: src/offline/NetworkStatus.fs
module Offline.NetworkStatus

open Fable.Core
open Fable.Core.JsInterop
open Browser

/// Access to global navigator object
[<Emit("navigator")>]
let private navigator : obj = jsNative

/// Check if currently online
let isOnline () : bool =
    navigator?onLine |> unbox<bool>

/// Subscribe to online/offline status changes
/// Returns a function to unsubscribe
let onStatusChange (callback: bool -> unit) : unit -> unit =
    let handleOnline = fun _ -> callback true
    let handleOffline = fun _ -> callback false

    window.addEventListener("online", handleOnline)
    window.addEventListener("offline", handleOffline)

    // Return cleanup function
    fun () ->
        window.removeEventListener("online", handleOnline)
        window.removeEventListener("offline", handleOffline)

/// Subscribe to visibility changes (for fallback sync trigger)
let onVisibilityChange (callback: bool -> unit) : unit -> unit =
    let handleVisibility = fun _ ->
        let isVisible = document?visibilityState = "visible"
        callback isVisible

    document.addEventListener("visibilitychange", handleVisibility)

    fun () ->
        document.removeEventListener("visibilitychange", handleVisibility)

File: src/offline/NetworkStatus.js

/**
 * Check if currently online
 */
export function isOnline() {
    return (navigator).onLine;
}

/**
 * Subscribe to online/offline status changes
 * Returns a function to unsubscribe
 */
export function onStatusChange(callback) {
    const handleOnline = (_arg) => {
        callback(true);
    };
    const handleOffline = (_arg_1) => {
        callback(false);
    };
    window.addEventListener("online", handleOnline);
    window.addEventListener("offline", handleOffline);
    return () => {
        window.removeEventListener("online", handleOnline);
        window.removeEventListener("offline", handleOffline);
    };
}

/**
 * Subscribe to visibility changes (for fallback sync trigger)
 */
export function onVisibilityChange(callback) {
    const handleVisibility = (_arg) => {
        const isVisible = document.visibilityState === "visible";
        callback(isVisible);
    };
    document.addEventListener("visibilitychange", handleVisibility);
    return () => {
        document.removeEventListener("visibilitychange", handleVisibility);
    };
}

